core/cost-model.md

```diff
@@ -0,0 +1,176 @@
+# Cost Model
+
+## Cost Model Constants
+
+- **TXN_SIGNATURE_COST**: 720  
+  Cost for each signature included in the transaction
+
+- **ED25519_SIGNATURE_COST**: 2400  
+  Cost for each precompile signature
+
+- **SECP256R1_SIGNATURE_COST**: 6690  
+  Cost for each precompile signature
+
+- **WRITABLE_ACCT_COST**: 300  
+  Cost for each writable account
+
+- **INSTR_DATA_COST**: 4  
+  Cost per byte of instruction data
+
+- **BUILTIN_INSTR_COST**: 3000  
+  Cost of built-in instructions
+
+- **NONBUILTIN_INSTR_COST**: 200000  
+  Cost of non-builtin instructions
+
+- **MAX_COMPUTE_UNIT_LIMIT**: 1400000  
+  Maximum compute unit limit
+
+- **HEAP_PAGE_COST**: 8  
+  Cost per heap page
+
+- **HEAP_PAGE_SIZE**: 32768  
+  Size of one heap page (32KB)
+
+- **MAX_LOADED_ACCT_DATA_SZ**: 64 * 1024 * 1024
+  Default size (64MB)
+
+- **SIMPLE_VOTE_COST**: 3428  
+  Fixed cost for simple vote transactions
+
+Before the release of feature gate
+`reserve_minimal_cus_for_builtin_instructions`
+(`C9oAhLxDBm3ssWtJx1yBGzPY55r2rArHmN1pbQn6HogH`) the following fixed
+costs were assigned to builtin instructions. After the feature gate is
+released, validators will assign `BUILTIN_INSTR_COST` to builtins and
+`NONBUILTIN_INSTR_COST` to non-builtins (non-builtings includes
+instructions that used to be builtins but were migrated to BPF).
+
+- **PRE_SIMD170_STAKE_PROGRAM_DEFAULT_COST**: 750
+- **PRE_SIMD170_CONFIG_PROGRAM_DEFAULT_COST**: 450
+- **PRE_SIMD170_ALT_PROGRAM_DEFAULT_COST**: 750
+- **PRE_SIMD170_SYSTEM_PROGRAM_DEFAULT_COST**: 150
+- **PRE_SIMD170_COMPUTE_BUDGET_PROGRAM_DEFAULT_COST**: 150
+- **PRE_SIMD170_BPF_LOADER_UPGRADEABLE_PROGRAM_DEFAULT_COST**: 2370
+- **PRE_SIMD170_BPF_LOADER_DEPRECATED_PROGRAM_DEFAULT_COST**: 1140
+- **PRE_SIMD170_BPF_LOADER_PROGRAM_DEFAULT_COST**: 570
+- **PRE_SIMD170_LOADER_V4_PROGRAM_DEFAULT_COST**: 2000
+
+## Messages
+
+### SignatureCost
+
+`SignatureCost` is the cost due to signature verification via builtin verifiers. This cost can be computed deterministically from the transaction payload.
+
+**`SignatureCost` is computed as:**
+```
+(TXN_SIGNATURE_COST * txn_sig_count) +
+(ED25519_SIGNATURE_COST * ed25519_sig_count) +
+(SECP256R1_SIGNATURE_COST * secp256r1_sig_count)
+```
```

Comment:
signature cost is made up from *4* parts: transaction signature, ed25519 (two versions, `strict` one behind feature gate), secp256k1, and sec256r1 (which is behind feature gate)